examples/common/convert_c4.py

```diff
@@ -4,105 +4,353 @@
 """C4 streaming dataset conversion scripts."""
 import os
 import platform
-from argparse import ArgumentParser, Namespace
-from typing import Dict, Iterable
+import warnings
+from argparse import Action, ArgumentError, ArgumentParser, Namespace
+from enum import Enum
+from textwrap import dedent
+from typing import Dict, Iterable, Optional
 
 import datasets as hf_datasets
+import numpy as np
 from streaming import MDSWriter
 from torch.utils.data import DataLoader, IterableDataset, get_worker_info
 from tqdm import tqdm
+from transformers import AutoTokenizer, PreTrainedTokenizerBase
+
+
+class ConcatMode(Enum):
+    NO_CONCAT = 'NO_CONCAT'
+    CONCAT_TEXT = 'CONCAT_TEXT'
+    CONCAT_TOKENS = 'CONCAT_TOKENS'
+
+
+class EmptyOrNonexistentDirectory(Action):
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        prospective_dir = values
+        if os.path.isdir(prospective_dir) and len(
+                os.listdir(prospective_dir)) > 0:
+            raise ArgumentError(self,
+                                f'--out_root={prospective_dir} must be empty')
+        else:
+            setattr(namespace, 'out_root', prospective_dir)
 
 
 def parse_args() -> Namespace:
     """Parse commandline arguments."""
-    args = ArgumentParser()
-    args.add_argument('--out_root', type=str, required=True)
-    args.add_argument('--compression', type=str, default=None)
-    args.add_argument('--splits',
-                      nargs='+',
-                      default=['train', 'train_small', 'val'])
+    parser = ArgumentParser(
+        description=
+        'Convert C4 into MDS format, optionally concatenating and tokenizing')
+    parser.add_argument('--out_root',
+                        type=str,
+                        required=True,
+                        action=EmptyOrNonexistentDirectory)
+    parser.add_argument('--compression', type=str, default=None)
+    parser.add_argument('--splits',
+                        nargs='+',
+                        default=['train', 'train_small', 'val'])
+
+    group = parser.add_mutually_exclusive_group(required=False)
+    group.add_argument('--concat_text',
+                       type=int,
+                       help='Number of characters to concatenate to')
+    group.add_argument(
+        '--concat_tokens',
+        type=int,
+        help='Convert text to tokens and concatenate up to this many tokens')
+
+    parser.add_argument('--tokenizer', type=str, required=False, default=None)
+    parser.add_argument('--bos_text', type=str, required=False, default=None)
+    parser.add_argument('--eos_text', type=str, required=False, default=None)
+    parser.add_argument('--wrap', default=False, action='store_true')
+
+    parsed = parser.parse_args()
+
+    if hasattr(parsed, 'concat_text') or hasattr(parsed, 'concat_tokens'):
```

Comment:
Why do the following two if statements need to be guarded by this if statement?

Reply:
it doesn't. removed.

Reply:
at one point I was using a subparser for `concat` and so it was possible for these to not be attributes of `parsed`. They always exist now so I will remove this guard